src/Analysis/Ast/Impl/Types/PythonClassType.cs

```diff
@@ -78,36 +77,38 @@ internal PythonClassType(string name, Location location)
         }
 
         public override IMember GetMember(string name) {
-            // Push/Pop should be lock protected.
-            if (Members.TryGetValue(name, out var member)) {
-                return member;
-            }
+            lock (_lock) {
```

Comment:
This seems heavy handed. What condition is this fixing?

@MikhailArkhipov Is this alright?

Reply:
Without it, multiple threads can pop the stack which gave the stack empty exception from before. I believe the reason we did not previously have to lock it is because the `Push` `Pop` mechanism from before was much simpler and just set _processing to null.

Reply:
Is this the case when class is generic and creation of specific type is being requested from multiple threads and the creation process is asking for the class members? In this case it should be sufficient to protect specific type creation.

Reply:
Also, push is down below, total lock is not really necessary, it may only be needed for the MRO access. Can we add a trace which thread re-enters and Debug.Fail on it? Such as simple variable with owning thread id (`Thread.CurrentThread.ManagedThreadId`) that checks if another thread reenters and asserts so you can break into debugger and see it?

Reply:
I don't think so. I think this is just because I switched out the Push and Pop from before to the new reentrant guard. I can check this directly off master and see if that's the case as well.

Reply:
I am trying to understand what is the case how you end up with empty stack. Even if there are 2 threads doing push1-push2-pop1-pop2 you should not get any exceptions. However, protecting stack inside the reentrancy guard on push and pop may be a good idea since `Stack<T>` is not a thread safe collection.

Reply:
I will work on debugging this and get back to you.

Reply:
I tried replacing Stack with a ConcurrentStack and removing the locks and there are no issues anymore. I believe what could be happening is that the stack is being pushed to at the same time by two threads and then popped twice? Whatever the case it has to do with how Stack is not thread safe. I'll make the scope of the locks tighter.

Reply:
Yeah that's what I thought. So you won't need global lock on `GetMember`. `ConcurrentStack` is probably an overkill, like `ConcurrentDictionary` it is generally for more advanced cases such as large stacks with heavy read/write. just add simple `lock()` around the stack operations. `lock(this)` is probably fine, its a small object with a short lock.

Reply:
Ok, sure I can lock just this.